array_api_compat/_internal.py

```diff
@@ -41,3 +42,31 @@ def wrapped_f(*args, **kwargs):
         return wrapped_f
 
     return inner
+
+
+def _get_all_public_members(module, exclude=None, extend_all=False):
+    """Get all public members of a module.
+    
+    Parameters
+    ----------
+    module : module
+        The module to get members from.
+    exclude : callable, optional
+        A callable that takes a name and returns True if the name should be
+        excluded from the list of members.
+    extend_all : bool, optional
```

Comment:
This is monkeypatching `torch.__all__` etc.? We don't want to do that.

Reply:
Yes, but this just keeps the current behavior. Take a look at https://github.com/data-apis/array-api-compat/blob/main/array_api_compat/torch/__init__.py#L3 

I have not checked whether this is still necessary, but probably we have to keep it this way?

Reply:
I checked with the following code:
```python
import torch

torch_all = set(torch.__all__)
public = set([name for name in dir(torch) if not name.startswith("_")])

print(torch_all.difference(public))
print(public.difference(torch_all))
```
And this gives:
```
set()
{'complex64', 'eig', 'special', ... , 'QInt8Storage', 'segment_reduce', 'ComplexDoubleStorage'}
```
So indeed `__all__` does not contain multiple members and most importantly it does not contain the dtypes.

Reply:
> Yes, but this just keeps the current behavior. Take a look at [main/array_api_compat/torch/__init__.py#L3](https://github.com/data-apis/array-api-compat/blob/main/array_api_compat/torch/__init__.py?rgh-link-date=2024-01-26T23%3A12%3A40Z#L3)

That code does not modify the `torch.__all__` list:

```py
>>> import torch
>>> torch_all = list(torch.__all__)
>>> import array_api_compat.torch
>>> torch_all2 = list(torch.__all__)
>>> torch_all == torch_all2
True
```

Generally speaking, this package should not monkeypatch the underlying libraries. 

> So indeed __all__ does not contain multiple members and most importantly it does not contain the dtypes.

Yes, that's a known issue. https://github.com/pytorch/pytorch/issues/91908

Reply:
Oh, now I understand. I did not mean to actually modify `torch.__all__` in place but copy and extend instead. I'll fix that behavior.